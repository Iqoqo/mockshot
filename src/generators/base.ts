import { SourceFile } from "ts-simple-ast";

export abstract class MockGenerator {
  abstract generate(fileDeclaration: SourceFile, snapshots: object): void;
  abstract getFilename(): string;
}
