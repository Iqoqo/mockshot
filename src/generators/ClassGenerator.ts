import { CodeBlockWriter, SourceFile } from "ts-simple-ast";
import pretty from "json-pretty";
import { MockGenerator } from "./base";

export class ClassGenerator extends MockGenerator {
  private mockDef: any = {};

  getFilename() {
    return "ClassMocks.ts";
  }

  generate(fileDecleration: SourceFile, snapshots: object) {
    const keys = Object.keys(snapshots);

    keys.forEach(key => this.parseSingleMock(snapshots[key]));

    const classNames = Object.keys(this.mockDef);

    classNames.forEach(className => {
      const mockClassName = className + "Mocks";

      const myClassFile = fileDecleration;

      const classDeclaration = myClassFile.addClass({
        name: mockClassName
      });

      classDeclaration.setIsExported(true);

      const methods = this.mockDef[className];

      const methodNames = Object.keys(methods);

      methodNames.forEach(methodName => {
        const mocks = methods[methodName];
        const mockNames = Object.keys(mocks);
        let mockTypes = mockNames.map(value => {
          return `"${value}"`;
        });
        const types = mockTypes.join(" | ");

        classDeclaration.addProperty({
          isStatic: true,
          name: methodName
        });

        const method = classDeclaration.addMethod({
          isStatic: true,
          parameters: [{ name: "mock", type: types }],
          name: methodName,
          returnType: "any"
        });

        method.setBodyText(writer =>
          writer.write("switch (mock)").block(() => {
            mockNames.forEach(mockName => {
              writer.write(`case "${mockName}":`).indentBlock(() => {
                writer.write(`return ${pretty(mocks[mockName])}`);
              });
            });
            writer.write(`default:`).indentBlock(() => {
              writer.write(`throw Error("Unknown mock: "+mock);`);
            });
          })
        );
      });
    });
  }

  private parseSingleMock(snap) {
    const snapshot = JSON.parse(snap);
    let classDef = this.mockDef[snapshot.className];

    if (!classDef) {
      classDef = {};
      this.mockDef[snapshot.className] = classDef;
    }

    let methodDef = classDef[snapshot.methodName];
    if (!methodDef) {
      methodDef = {};
      classDef[snapshot.methodName] = methodDef;
    }

    let mockName = methodDef[snapshot.mockName];
    if (!mockName) {
      methodDef[snapshot.mockName] = snapshot.mock;
    } else {
      throw Error(
        "Duplicate mock name for method " +
          snapshot.methodName +
          ": " +
          snapshot.mockName
      );
    }
  }
}
