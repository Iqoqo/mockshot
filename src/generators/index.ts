export * from "./ApiGenerator";
export * from "./ClassGenerator";
export * from "./SpyGenerator";
