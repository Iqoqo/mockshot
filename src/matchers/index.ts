import stringify from 'json-stable-stringify';
import { toMatchApiMock } from "./ApiMockMatcher";
import { toMatchClassMock } from "./ClassMockMatcher";

declare global {
    namespace jest {
        // tslint:disable-next-line:interface-name
        interface Matchers<R> {
            toMatchClassMock(
                className: string,
                methodName: string,
                mockName: string,
                ignoredKeyPaths?: string[]
            ): R;
            toMatchApiMock(
                response: object,
                mockName: string,
            ): R;
        }
    }
}

expect.addSnapshotSerializer({
    test: val => val.mock,
    print: val => stringify(val, { space: '  ' })
});

const toMatchMock = toMatchClassMock

expect.extend({ toMatchMock, toMatchClassMock, toMatchApiMock });

export { toMatchMock, toMatchClassMock, toMatchApiMock };
