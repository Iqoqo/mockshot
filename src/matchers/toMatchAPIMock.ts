import { cloneDeep, get, set } from "lodash";
import pretty from "json-pretty";
import { generateMocks } from "../generateMocks";

export enum HttpMethods {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  PATCH = "PATCH",
  DELETE = "DELETE"
}

export enum RequestModules {
  request = "request",
  axios = "axios"
}

export interface IApiSnapshot {
  mockName: string;
  url: string;
  httpMethod: "post" | "get" | "put" | "delete" | "patch";
  mock: {
    statusCode: number;
    body: any;
    error?: string;
  };
}

export const ApiSnapshotTag = "[APISnap]";

expect.addSnapshotSerializer({
  test: val => val.mock,
  print: val => pretty(val)
});

function getSnapshotName(testName, snapshotTag: string) {
  // TODO: If there are more the 1 snapshots (indicated by the number at the end) throw an exception
  return `${testName}: ${snapshotTag} 1`;
}

function getSnapshotTag(
  methodName: HttpMethods,
  url: string,
  mockName: string
) {
  return `[mockshot] [[${methodName} ${url} ${mockName}]]`;
}

let commonSnapshotState;

afterAll(async () => {
  if (
    commonSnapshotState &&
    (commonSnapshotState.added > 0 || commonSnapshotState.updated > 0)
  ) {
    await generateMocks();
  }
});

function toMatchApiMock(received, returnValue: string) {
  commonSnapshotState = this.snapshotState;

  const snapshotTag = getSnapshotTag(
    received.config.method.toUpperCase(),
    received.url,
    returnValue
  );
  const snapshotName = getSnapshotName(this.currentTestName, snapshotTag);
  const currentSnapshot = this.snapshotState._snapshotData[snapshotName];

  const snapshot = {
    httpMethod: received.config.method.toUpperCase(),
    url: received.config.url,
    mockName: returnValue,
    mock: { statusCode: received.status, body: received.data }
  };

  const snapshotNameTag = `[${received.config.method.toUpperCase()} ${
    received.config.url
  } ${returnValue}]`;

  const result = expect(snapshot).toMatchSnapshot(
    `[mockshot] ${ApiSnapshotTag} [${snapshotNameTag}]`
  );

  const pass = result === undefined;

  return { pass };
}

expect.extend({ toMatchApiMock });

export { toMatchApiMock };
