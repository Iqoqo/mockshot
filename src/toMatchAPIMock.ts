import pretty from "json-pretty";
import { generateMocks } from "./generateMocks";

export enum HttpMethods {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  PATCH = "PATCH",
  DELETE = "DELETE"
}

export const RequestModules = {
  request: "request",
  axios: "axios",
  r2: "r2",
  fetch: "fetch"
}

expect.addSnapshotSerializer({
  test: val => val.mock,
  print: val => pretty(val)
});

let commonSnapshotState;

afterAll(async () => {
  if (
    commonSnapshotState &&
    (commonSnapshotState.added > 0 || commonSnapshotState.updated > 0)
  ) {
    await generateMocks();
  }
});

async function toMatchApiMock(
  received,
  requestModules,
  returnValue: string
) {
  commonSnapshotState = this.snapshotState;

  let url;
  let method;
  let mock;

  switch (requestModules) {
    case RequestModules.r2:
      const res = await received.response;
      method = received.opts.method.toUpperCase();
      url = res.url;
      mock = { status: res.status, body: res.body }
      break;
    case RequestModules.axios:
    default:
      method = received.config.method.toUpperCase();
      url = received.url;
      mock = { status: received.status, body: received.data }
      break;
  }

  const snapshot = {
    method,
    url,
    mockName: returnValue,
    mock
  };

  const snapshotNameTag = `[${method} ${url} ${returnValue}]`;
  const result = expect(snapshot).toMatchSnapshot(
    `[mockshot] [${snapshotNameTag}]`
  );

  const pass = result === undefined;
  return { pass };
}

expect.extend({ toMatchApiMock });

export { toMatchApiMock };
