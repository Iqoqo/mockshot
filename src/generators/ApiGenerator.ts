import { CodeBlockWriter, SourceFile } from "ts-simple-ast";
import util from "util";
import _ from "lodash";

import { MockGenerator } from "./base";
import { ApiSnapshotTag, IApiSnapshot } from "../matchers/contracts";

const methodParameter = "url";

export class ApiGenerator extends MockGenerator {
  getFilename() {
    return "API.ts";
  }

  filterSnapKeys(keys: string[]): string[] {
    return keys.filter(key => _.includes(key, ApiSnapshotTag));
  }

  generate(fileDeclaration: SourceFile, snapshots: object) {
    const parsed = this.parse(snapshots);

    fileDeclaration
      .addClass({ name: "API" })
      .setIsExported(true)
      .addMethods(this.getMethodsFrom(parsed))
      .map(methodDeclaration => {
        methodDeclaration.setBodyText(
          this.getSwitchStatement(
            methodParameter,
            parsed[methodDeclaration.getName()]
          )
        );
      });

    fileDeclaration.addInterfaces(this.getInterfacesFrom(parsed));
  }

  private getMethodsFrom(parsed) {
    return Object.keys(parsed).map(method => ({
      isStatic: true,
      parameters: [{ name: methodParameter, type: "T" }],
      name: `${method}<T extends keyof ${method}Responses>`,
      returnType: `${method}Responses[T]`
    }));
  }

  private getInterfacesFrom(parsed) {
    return Object.keys(parsed).map(method => ({
      name: `${method}Responses`,
      isExported: true,
      properties: Object.keys(parsed[method]).map(url => {
        const types = Object.keys(parsed[method][url])
          .map(type => `"${type}": any`)
          .join();
        return { name: `"${url}"`, type: `{${types}}` };
      })
    }));
  }

  private parsed = {};

  private validate(snap, key) {
    if (!this.isHttpMethodValid(snap.httpMethod)) {
      throw Error(
        `Invalid http method '${snap.httpMethod}' in snapshot '${key}'`
      );
    }
    if (!snap.url) {
      throw Error(`Missing property url in snapshot '${key}'`);
    }
    if (!snap.mockName) {
      throw Error(`Missing property mockName in snapshot '${key}'`);
    }
    if (!snap.mock) {
      throw Error(`Missing property mock in snapshot '${key}'`);
    }
    if (!snap.mock.statusCode) {
      throw Error(`Missing property statusCode in snapshot '${key}'`);
    }
  }

  private fillMissingPath(snap) {
    if (!_.has(this.parsed, snap.httpMethod)) {
      this.parsed[snap.httpMethod] = {};
    }
    if (!_.has(this.parsed[snap.httpMethod], snap.url)) {
      this.parsed[snap.httpMethod][snap.url] = {};
    }
  }

  private parse(snapshots: object) {
    _.keys(snapshots).forEach(key => {
      const snap: IApiSnapshot = snapshots[key];

      this.validate(snap, key)
      this.fillMissingPath(snap)

      if (_.has(this.parsed[snap.httpMethod][snap.url], snap.mockName)) {
        throw Error(
          `Snapshot duplication: snapshot with the same httpMethod, URL and mockName already exists '${key}'`
        );
      }
      this.parsed[snap.httpMethod][snap.url][snap.mockName] = snap.mock;
    });

    return this.parsed;
  }

  private isHttpMethodValid(method: string): boolean {
    return _.includes(["post", "get", "put", "delete", "patch"], method);
  }

  private getSwitchStatement(methodParameter: string, options: object) {
    return (writer: CodeBlockWriter) =>
      writer
        .write(`switch (${methodParameter})`)
        .block(() =>
          Object.keys(options).map(key =>
            writer
              .write(`case "${key}":`)
              .indentBlock(() =>
                writer.write(
                  `return ${util.inspect(options[key], { depth: null })}`
                )
              )
          )
        );
  }
}
