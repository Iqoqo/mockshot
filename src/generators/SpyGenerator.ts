/**
 * Copyright (c) Samsung, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import stringify from "json-stable-stringify";
import path from "path";
import { SourceFile, VariableDeclarationKind } from "ts-simple-ast";
import { IClassSnapshot, ISnapshot, SingleClassMockTree } from "../contracts";
import { ClassSnapshotTag } from "../matchers/ClassMockMatcher";
import { MockGenerator } from "./base";
import { parseClassSnapshots } from "./parsers";

export class ClassSpyGenerator extends MockGenerator {
  generate(
    getFile: (filename: string) => SourceFile,
    allSnapshots: ISnapshot[]
  ) {
    const classSnapshots = this.filterSnapshots(allSnapshots);
    const tree = parseClassSnapshots(classSnapshots);

    Object.keys(tree).forEach(filePath => {
      const file = getFile(this.getFilePath(filePath));
      const { className, classContent } = tree[filePath];
      this.writeSpy(file, className, classContent);
    });
  }

  private filterSnapshots(allSnapshots: ISnapshot[]): IClassSnapshot[] {
    return allSnapshots.filter(snap => snap.key.includes(ClassSnapshotTag));
  }

  private writeSpy(
    file: SourceFile,
    className: string,
    classTree: SingleClassMockTree
  ) {
    this.writeMockshotMockInterface(file);
    this.writeSpyInterface(file, className, classTree);
    this.writeClassTree(file, classTree);
    this.writeGetSpyFunction(file);
    this.writeClassSpy(file, className, classTree);
  }

  private writeMockshotMockInterface(file: SourceFile) {
    file.addInterface({
      name: "MockshotMock<P, T = {}>",
      extends: ["jest.Mock"],
      properties: [
        { name: "mockImplementation(mockName: P)", type: "jest.Mock<T>" },
        {
          name: "mockImplementation(fn: (...args: any[]) => any)",
          type: "jest.Mock<T>"
        },
        { name: "mockImplementationOnce(mockName: P)", type: "jest.Mock<T>" },
        {
          name: "mockImplementationOnce(fn: (...args: any[]) => any)",
          type: "jest.Mock<T>"
        },
        {
          name: "getMock(mockName: P)",
          type: "any"
        }
      ]
    });
  }

  private writeSpyInterface(
    file: SourceFile,
    className: string,
    classTree: SingleClassMockTree
  ) {
    file.addInterface({
      name: `${className}Spy`,
      properties: this.getMethodsTypes(classTree)
    });
  }

  private writeClassTree(file: SourceFile, classTree: SingleClassMockTree) {
    file.addVariableStatement({
      declarationKind: VariableDeclarationKind.Const,
      declarations: [{ name: "classTree", initializer: stringify(classTree) }]
    });
  }

  private writeGetSpyFunction(file: SourceFile) {
    file.addFunction({
      name: "getSpy<P extends string>",
      parameters: [{ name: "methodName", type: "string" }],
      returnType: "MockshotMock<P>",
      bodyText: writer => {
        writer.writeLine("const newSpy = jest.fn() as MockshotMock<P>;");
        writer.writeLine(
          "const getMock = mockName => classTree[methodName][mockName].mock;"
        );
        writer.writeLine(
          "const { mockImplementation, mockImplementationOnce } = newSpy;"
        );
        writer.writeLine("newSpy.mockImplementation = mockName =>");
        writer.writeLine('  typeof mockName === "string"');
        writer.writeLine("    ? mockImplementation(() => getMock(mockName))");
        writer.writeLine("    : mockImplementation(mockName);");
        writer.writeLine("newSpy.mockImplementationOnce = mockName =>");
        writer.writeLine('  typeof mockName === "string"');
        writer.writeLine(
          "    ? mockImplementationOnce(() => getMock(mockName))"
        );
        writer.writeLine("    : mockImplementationOnce(mockName);");
        writer.writeLine("newSpy.getMock = getMock;")
        writer.writeLine("return newSpy;");
      }
    });
  }

  private writeClassSpy(
    file: SourceFile,
    className: string,
    classTree: SingleClassMockTree
  ) {
    file.addVariableStatement({
      declarationKind: VariableDeclarationKind.Const,
      isExported: true,
      declarations: [
        {
          name: this.getSpyClassName(className),
          type: this.getSpyInterfaceName(className),
          initializer: writer => {
            writer.writeLine("{");
            Object.keys(classTree).map(methodName =>
              writer.writeLine(`    ${methodName}: getSpy("${methodName}"),`)
            );
            writer.writeLine("}");
          }
        }
      ]
    });
  }

  private getMethodsTypes(classTree: SingleClassMockTree) {
    return Object.keys(classTree).map(methodName => {
      const methodTree = classTree[methodName];
      const methodArgUnionType = this.getMethodArgUnionType(methodTree);
      return { name: methodName, type: `MockshotMock<${methodArgUnionType}>` };
    });
  }

  private getMethodArgUnionType(methodTree: {
    [mockName: string]: any;
  }): string {
    return Object.keys(methodTree)
      .map(mockName => `"${mockName}"`)
      .join(" | ");
  }

  private getSpyInterfaceName(className: string): string {
    return className + "Spy";
  }

  private getSpyClassName(className: string): string {
    return className.charAt(0).toLowerCase() + className.slice(1) + "Spy";
  }

  private getFilePath(filePath): string {
    return path.join("spies", filePath);
  }
}
