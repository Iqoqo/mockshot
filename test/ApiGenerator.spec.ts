import Project, { SourceFile } from "ts-simple-ast";
import { ApiGenerator } from "../src/generators/ApiGenerator";
import { IApiSnapshot, ApiSnapshotTag } from "../src/matchers/toMatchAPIMock";

const snapshots = {
  [`${ApiSnapshotTag} 1`]: {
    url: "/hello/world",
    httpMethod: "post",
    mock: {
      statusCode: 200,
      body: {
        foo: {
          bar: "123"
        }
      }
    },
    mockName: "success"
  },
  [`${ApiSnapshotTag} 2`]: {
    url: "/hello/world",
    httpMethod: "post",
    mock: {
      statusCode: 501,
      error: "not implemented"
    },
    mockName: "fail"
  },
  [`${ApiSnapshotTag} 3`]: {
    url: "/bye/world",
    httpMethod: "get",
    mock: {
      body: { foo: { bar: "123" } },
      statusCode: 200
    },
    mockName: "success"
  },
  [`${ApiSnapshotTag} 4`]: {
    url: "/hello/world",
    httpMethod: "get",
    mock: {
      statusCode: 501,
      error: "not implemented!"
    },
    mockName: "fail"
  }
} as { [key: string]: IApiSnapshot };

const snapshotsMissingName = {
  ...snapshots,
  [`${ApiSnapshotTag} 5`]: {
    url: "/hello/world/2",
    httpMethod: "get",
    mock: {
      statusCode: 501,
      error: "not implemented!"
    }
  }
};

const snapshotsMissingUrl = {
  ...snapshots,
  [`${ApiSnapshotTag} 5`]: {
    httpMethod: "get",
    mock: {
      statusCode: 501,
      error: "not implemented!"
    },
    mockName: "fail"
  }
};

const snapshotsInvalidMethod = {
  ...snapshots,
  [`${ApiSnapshotTag} 5`]: {
    url: "/hello/world/2",
    httpMethod: "take",
    mock: {
      statusCode: 501,
      error: "not implemented!"
    },
    mockName: "fail"
  }
};

const snapshotsMissingStatus = {
  ...snapshots,
  [`${ApiSnapshotTag} 5`]: {
    url: "/hello/world/2",
    httpMethod: "get",
    mock: {
      error: "not implemented!"
    },
    mockName: "fail"
  }
};

const snapshotsMissingMock = {
  ...snapshots,
  [`${ApiSnapshotTag} 5`]: {
    url: "/hello/world/2",
    httpMethod: "get",
    mockName: "fail"
  }
};

const snapshotsDuplication = {
  ...snapshots,
  [`${ApiSnapshotTag} dup`]: snapshots[Object.keys(snapshots)[0]]
};

describe("ApiGenerator", () => {
  let fileDeclaration: SourceFile;
  let generator: ApiGenerator;

  beforeEach(() => {
    const project = new Project();
    fileDeclaration = project.createSourceFile("api");
    generator = new ApiGenerator();
  });

  it("should throw if snapshot is missing mockName", async () => {
    expect(
      await generator.generate(fileDeclaration, snapshotsMissingName)
    ).toThrowError();
  });

  it("should throw if snapshot is missing url", async () => {
    expect(
      await generator.generate(fileDeclaration, snapshotsMissingUrl)
    ).toThrowError();
  });

  it("should throw if snapshot has invalid HTTP method", async () => {
    expect(
      await generator.generate(fileDeclaration, snapshotsInvalidMethod)
    ).toThrowError();
  });

  it("should throw if snapshot is missing status code", async () => {
    expect(
      await generator.generate(fileDeclaration, snapshotsMissingStatus)
    ).toThrowError();
  });

  it("should throw if snapshot is missing mock property", async () => {
    expect(
      await generator.generate(fileDeclaration, snapshotsMissingMock)
    ).toThrowError();
  });

  it("should throw if snapshot has duplicate mocks", async () => {
    expect(
      await generator.generate(fileDeclaration, snapshotsDuplication)
    ).toThrowError();
  });

  it("should generate API endpoint mock", async () => {
    await generator.generate(fileDeclaration, snapshots);

    expect(fileDeclaration._compilerNode.text).toMatchSnapshot(
      "[API] [generator]"
    );
  });
});
