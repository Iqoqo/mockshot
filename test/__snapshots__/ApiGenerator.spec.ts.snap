// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiGenerator Should generate API endpoint mock: [API] [generator] 1`] = `
"export class API {
    static post<T extends keyof postResponses>(url: T): postResponses[T] {
        switch (url) {
            case \\"/hello/world\\":
                return { success: { statusCode: 200, body: { foo: { bar: '123' } } },
                  fail: { statusCode: 501, error: 'not implemented' } }
        }
    }

    static get<T extends keyof getResponses>(url: T): getResponses[T] {
        switch (url) {
            case \\"/bye/world\\":
                return { success: { body: { foo: { bar: '123' } }, statusCode: 200 } }
            case \\"/hello/world\\":
                return { fail: { statusCode: 501, error: 'not implemented!' } }
        }
    }
}

export interface postResponses {
    \\"/hello/world\\": {success: any,fail: any};
}

export interface getResponses {
    \\"/bye/world\\": {success: any};
    \\"/hello/world\\": {fail: any};
}
"
`;

exports[`ApiGenerator should generate API endpoint mock: [API] [generator] 1`] = `
"export class API {
    static post<T extends keyof postResponses>(url: T): postResponses[T] {
        switch (url) {
            case \\"/hello/world\\":
                return { success: { statusCode: 200, body: { foo: { bar: '123' } } },
                  fail: { statusCode: 501, error: 'not implemented' } }
        }
    }

    static get<T extends keyof getResponses>(url: T): getResponses[T] {
        switch (url) {
            case \\"/bye/world\\":
                return { success: { body: { foo: { bar: '123' } }, statusCode: 200 } }
            case \\"/hello/world\\":
                return { fail: { statusCode: 501, error: 'not implemented!' } }
        }
    }
}

export interface postResponses {
    \\"/hello/world\\": {success: any,fail: any};
}

export interface getResponses {
    \\"/bye/world\\": {success: any};
    \\"/hello/world\\": {fail: any};
}
"
`;
